nodecg-io-core/extension/serviceBundle.ts

```diff
@@ -0,0 +1,75 @@
+import { NodeCGIOCore } from ".";
+import { NodeCG } from "nodecg/types/server";
+import { Service, ServiceProvider } from "./types";
+import { emptySuccess, Result } from "./utils/result";
+
+import * as fs from "fs";
+import * as path from "path";
+
+/**
+ * Class helping to create a nodecg-io service
+ */
+export class ServiceBundle {
+    private core: NodeCGIOCore | undefined;
+    private service: Service<unknown, unknown>;
+    private nodecg: NodeCG;
+
+    constructor(nodecg: NodeCG, serviceName: string, ...pathSegments: string[]) {
+        this.nodecg = nodecg;
+        this.service = {
+            schema: this.readSchema(...pathSegments),
+            serviceType: serviceName,
+            validateConfig: this.validateConfig,
+            createClient: this.createClient(nodecg),
+            stopClient: this.stopClient,
+        };
+        nodecg.log.info(this.service.serviceType + " bundle started");
+        this.core = (nodecg.extensions["nodecg-io-core"] as unknown) as NodeCGIOCore | undefined;
+        if (this.core === undefined) {
+            this.nodecg.log.error(
+                "nodecg-io-core isn't loaded! " + this.service.serviceType + " bundle won't function without it.",
+            );
+        }
+    }
+
+    public register(): ServiceProvider<any> | undefined {
+        if (this.core === undefined) {
+            return undefined;
+        } else {
+            return this.core.registerService(this.service);
+        }
+    }
+
+    async validateConfig(config: unknown): Promise<Result<void>> {
+        this.nodecg.log.error(
+            this.service.serviceType +
+                " has not provided a validateConfig method! The service won't function without it.",
+        );
+        return emptySuccess();
+    }
+
+    createClient(nodecg: NodeCG): (config: unknown) => Promise<Result<unknown>> {
```

Comment:
Returning a function was the solution to getting access to the nodecg instance when it was a loose function.
This is not needed here because the function can access the nodecg instance through the local variable (`this.nodecg`).
Therefore we can match the function signature with the service interface and can extend it.
Also please use [generics](https://www.typescriptlang.org/docs/handbook/generics.html#generic-classes) to ensure type safety.
```suggestion
    async createClient(config: unknown): Promise<Result<unknown>> {
```